pkg/server/container_execsync.go

```diff
@@ -140,40 +131,33 @@ func (c *criContainerdService) execInContainer(ctx context.Context, id string, o
 		}
 	})
 
-	resCh := make(chan execResult, 1)
-	go func() {
-		// Wait will return if context is cancelled.
-		exitCode, err := process.Wait(ctx)
-		resCh <- execResult{
-			exitCode: exitCode,
-			err:      err,
-		}
-		glog.V(2).Infof("Exec process %q exits with exit code %d and error %v", execID, exitCode, err)
-	}()
+	exitCh, err := process.Wait(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait for process %v", err)
+	}
 	if err := process.Start(ctx); err != nil {
 		return nil, fmt.Errorf("failed to start exec %q: %v", execID, err)
 	}
 
-	var timeoutCh <-chan time.Time
-	if opts.timeout == 0 {
-		// Do not set timeout if it's 0.
-		timeoutCh = make(chan time.Time)
-	} else {
-		timeoutCh = time.After(opts.timeout)
+	if opts.timeout != 0 {
+		ctx, cancel = context.WithTimeout(ctx, opts.timeout)
```

Comment:
We should not use context timeout here. Once the context is cancelled, you'll not even be able to call `Delete`. It means that once there is a timeout, there will be resource leakage, because we won't be able to delete the process.

Although we could pass `context.Background()` to `Delete`, I prefer keeping the original timeout mechanism. :)